### PR TITLE
USR-243 fix(clerk-js): Handle the construction of zxcvbn errors with informat…

### PR DESCRIPTION
**Pull request by raptisj**
…ion from FAPI

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [x] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [x] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [x] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

Handle the construction of zxcvbn errors with information from FAPI

https://github.com/clerkinc/javascript/assets/23478420/5905e518-0ef9-4548-a615-3176886c9c8c

### CODE REVIEW

File: packages/types/src/api.ts [hunk 1]
```diff
@@ -9,6 +9,12 @@ export interface ClerkAPIError {
     paramName?: string;
     sessionId?: string;
     emailAddresses?: string[];
+    zxcvbn?: {
+      suggestions: {
+        code: string;
+        message: string;
```
Comment (panteliselef) [inline packages/types/src/api.ts:15]: Seems like we are not using this in the codebase, right ?

File: packages/clerk-js/src/ui/utils/passwordUtils.ts [hunk 2]
```diff
@@ -29,13 +29,29 @@ type LocalizationConfigProps = {
...
 export const createPasswordError = (errors: ClerkAPIError[], localizationConfig: LocalizationConfigProps) => {
   if (!localizationConfig) {
     return errors[0].longMessage;
   }
 
   const { t, locale, passwordSettings } = localizationConfig;
 
+  if (errors.length && errors[0] && errors[0].code === 'form_password_not_strong_enough') {
+    const message = errors[0].meta?.zxcvbn?.suggestions
+      ?.map((s: ZxcvbnSuggestionKeys) => {
```
Comment (panteliselef) [inline packages/clerk-js/src/ui/utils/passwordUtils.ts:47]: ❓Why do we need an explicit type here ? Should  this  be inferred based on the types you added bellow ?

**Comment by braden-clerk** [comment]
!snapshot

File: packages/clerk-js/src/ui/utils/passwordUtils.ts (hunk 2 above)
Comment (raptisj) [reply packages/clerk-js/src/ui/utils/passwordUtils.ts:47]: @panteliselef I agree this type doesn’t add much value

File: packages/types/src/json.ts [hunk 3]
```diff
@@ -273,6 +273,12 @@ export interface ClerkAPIErrorJSON {
     param_name?: string;
     session_id?: string;
     email_addresses?: string[];
+    zxcvbn?: {
+      suggestions: {
+        code: string;
+        message: string;
+      }[];
```
Comment (anagstef) [inline packages/types/src/json.ts:280]: ❓ Do you think we should extract this in a separate type?

File: packages/clerk-js/src/ui/utils/passwordUtils.ts [hunk 4]
```diff
@@ -29,13 +29,24 @@ type LocalizationConfigProps = {
...
 };
+
 export const createPasswordError = (errors: ClerkAPIError[], localizationConfig: LocalizationConfigProps) => {
   if (!localizationConfig) {
     return errors[0].longMessage;
   }
 
   const { t, locale, passwordSettings } = localizationConfig;
 
+  if (errors.length && errors[0] && errors[0].code === 'form_password_not_strong_enough') {
```
Comment (anagstef) [inline packages/clerk-js/src/ui/utils/passwordUtils.ts:40]:
```suggestion
  if (errors?.[0]?.code === 'form_password_not_strong_enough') {
```

File: packages/types/src/json.ts (hunk 3 above)
Comment (raptisj) [reply packages/types/src/json.ts:280]: @anagstef I believe being a bit more explicit here is no harm, so we can keep it as is.

**Review by nikosdouvlis — APPROVED** [review]

**Comment by clerk-cookie** [comment]
This PR has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.

**Pull request merged**